### PR TITLE
Add SQL planner support for EXISTS subqueries

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2219

 # Rationale for this change

Add SQL query planner support for EXISTS subqueries

# What changes are included in this PR?

The SQL query planner has been updated so that it passes information about the outer schema to subqueries, so that correlated subqueries can resolve fields from the outer query.

# Are there any user-facing changes?

Yes, SQL queries with EXISTS subqueries will now result in valid logical plan instead of an error.

### CODE REVIEW

**Comment by andygrove** [comment]
@alamb Here is the first PR to actually support subqueries in the SQL planner. I'm sure there will be follow up work to make sure the optimizer rules work correctly but this at least enables the creation of a logical plan from a SQL query containing EXISTS subqueries.

**Comment by alamb** [comment]
I plan to review this shortly

File: datafusion/core/src/sql/planner.rs [hunk 1]
```diff
@@ -443,16 +463,21 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         &self,
         t: TableWithJoins,
         ctes: &mut HashMap<String, LogicalPlan>,
+        outer_schema: Option<&DFSchema>,
```
Comment (alamb) [inline datafusion/core/src/sql/planner.rs:466]: Given that `outer` means something already in terms of joins, I wonder if a name like `outer_query_schema` might potentially cause less confusion

File: datafusion/core/src/sql/planner.rs [hunk 2]
```diff
@@ -4182,4 +4264,36 @@ mod tests {
...
+            (SELECT * FROM person \
+            WHERE last_name = p.last_name \
+            AND state = p.state)";
+
+        let expected = "Projection: #p.id\
+        \n  Filter: EXISTS (Subquery: Projection: #person.id, #person.first_name, #person.last_name, #person.age, #person.state, #person.salary, #person.birth_date, #person.😀\
+        \n  Filter: #person.last_name = #p.last_name AND #person.state = #p.state\
+        \n    TableScan: person projection=None)\
+        \n    SubqueryAlias: p\
+        \n      TableScan: person projection=None";
```
Comment (alamb) [inline datafusion/core/src/sql/planner.rs:4296]: I am not sure what this plan is doing (as it seems to have a `LogicalPlan::Filter` node that has multiple inputs which I didn't think was possible 🤔 )

With a correlated subquery like this I think the classic plan uses a `Semi-Join` like:

```
"Projection: #p.id\
  Join(type=Semi): #person.last_name = #p.last_name AND #person.state = #p.state\
    TableScan: person projection=None)\
    SubqueryAlias: p\
      TableScan: person projection=None";
```

**Review by alamb — COMMENTED** [review]
I don't think the plans are right -- I think we need to model them as Joins (or something more sophisticated than Filters with two inputs)

File: datafusion/core/src/sql/planner.rs (hunk 1 above)
Comment (andygrove) [reply datafusion/core/src/sql/planner.rs]: That makes sense. I renamed this.

File: datafusion/core/src/sql/planner.rs (hunk 2 above)
Comment (andygrove) [reply datafusion/core/src/sql/planner.rs]: The query is hard to parse here and I have pushed a change to improve the formatting in the test. There is one filter in the subquery and one filter in the outer query.

My plan was to implement an optimizer rule to handle re-writing to a semi-join. This rule would be applied regardless of whether the query was created by the SQL planner or via the DataFrame API. Does that make sense?

Comment (alamb) [reply datafusion/core/src/sql/planner.rs]: Thank you for the change. It is much easier to see what is going on now.

**Review by alamb — APPROVED** [review]
